### PR TITLE
Capstone Project: Budget App

### PR DESCRIPTION
**Pull request by arnoldnekemiah**
### **Welcome To this PR 🚢** 
_Below is my video presentation link;_

 Watch a live video demo of this project <a href="https://drive.google.com/file/d/1ok3SaP2PFFD174r5aPoNG_eTUAHxMezC/view?usp=sharing">here</a>

This is the link to the deployed site  [Live Demo Link](https://money-nm7c.onrender.com)

**I did the following;**
### Interactions
_**Splash screen**_

- A simple page with the name of your app (yes, you need to choose one), and links to the sign up and log in pages.
- Sign up and log in pages

- The user should be able to register in the app with full name, email and password (all mandatory).
- The user can log into the app using email and password.
- If the user is not logged in, they can't access pages that require the user to be logged in (all the pages described below).
- Home page (categories page)

When the user logs in, they are presented with the categories page.

- For each category, the user can see their name, icon and the total amount of all the transactions that belongs to that category.
- When the user clicks (or taps) on a category item, the application navigates to the transactions page for that category.
- There is a button "add a new category" at the bottom that brings the user to the page to create a new category.
- Transactions page

**_Category Page_**

- For a given category, the list of transactions is presented, ordered by the most recent.
- At the top of the page the user could see the total amount for the category (sum of all of the amounts of the transactions in that category).
- There is a button "add a new transaction" at the bottom that brings the user to the page to create a new transaction.
- When the user clicks on the "Back" button (<), the user navigates to the home page.
- "Add a new category" page

- The user fills out a form to create a _new category_, indicating their name and icon (both mandatory).
- The user clicks (or taps) the "Save" button to create the new category, and is taken to the home page on success.
- When the user clicks on the "Back" button (<), the user navigates to the home page.
- "Add a new transaction" page

_The user fills out a form to create a new transaction with:_

- name (mandatory)
- amount (mandatory)
- categories (mandatory at least one)
- The user click (or taps) the "Save" button to create the new transaction, and is taken to the transactions page for that category.
- When the user clicks on the "Back" button (<), the user navigates to the transactions page for that category.

### CODE REVIEW

File: app/views/groups/index.html.erb [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    <% @groups.each_with_index do |cat, index|%>
+      <%= link_to group_path(cat) do %>
+        <div class="category card mb-3">
+          <div class="row g-0">
+             <div class="col-md-4">
+             <%= image_tag(cat.icon, class: "img-fluid", alt: cat.name, style: "height: 100px; width:100px;") if cat.icon.attached? %>
+        </div>
+            <div class="col-md-8">
+              <div class="card-body text-dark">
+                <div class="d-flex justify-content-between">
```
Comment (Shedrack-Sunday) [inline app/views/groups/index.html.erb]: - [ ] Great job 👏 , however, let's stick to the design flow for the category as it is on the transaction which was a pattern provided.
- The top bar should have a blue background. The color used is `#3778c2`
- the "Add new category" should be a button at the bottom of the page. The color used is `#5fb523`

### Expected design:
image

### Your design:
image

File: app/views/spendings/index.html.erb [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+  <div class="mb-10">
+  <% if @spendings.empty?%>
+    <p class="text-center pt-16 text-xl">You don't have any transactions yet !!</p>
+  <% else %>
+    <ul class="flex flex-col justify-center pt-12 ">
+      <li class="w-[90%] mx-auto flex flex-col items-center gap-3 shadow-lg px-3 py-5  my-5 bg-white">
+          <div class="flex  justify-center items-center">
+              <div class="text-4xl me-4">
+                <%= image_tag(@group.icon, alt: "Group Icon", class: "img-fluid", style: "height: 100px; width:100px;") %>
+              </div>
```
Comment (Shedrack-Sunday) [inline app/views/spendings/index.html.erb]: - [ ] For the transaction page. Kindly stick to the color, structure and pattern which was provided in the behance design. The aim of this is to see how well you can replicate a design.

- The top bar should have a blue background. The color used is `#3778c2`
- The "Add new transaction" button should be at the bottom of the page and stretched horizontally as in the design. The color used is `#5fb523`
- Kindly fix the structure of each transaction to be placed in a flex horizontally 

### Expected design:
image

### Your design:
image

File: app/views/groups/index.html.erb [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+                  <h5 class="card-title"><%= cat.name %></h5>
+                <p class="card-text">Amount: $<%= @totals[index] || 0 %></p>
+                </div>
```
Comment (Shedrack-Sunday) [inline app/views/groups/index.html.erb]: - [ ] This section is meant to show the total amount of all the transactions that belong to that category, not the amount.
- For example, if you have 4 transaction, it shows 4, not the sum of the transactions

### Change view:
image

**Review by Shedrack-Sunday — CHANGES_REQUESTED** [review]
Hello @arnoldnekemiah       👋 

# CHANGES REQUIRED ♻️

Impressive work on the project so far👏
You've done extremely well so far in this project,  However, there are some adjustments you need to make to move to the next project. 💯



## Highlights   
- Good use of GitHub features ✔️   
- Nice work with the rails testing features ✔️
- Great work on the routes 👏 
- Kudus personalizing the design   ✔️  
- Awsome work with the database design 👏 

### Confirmed Test view passed: 💯 
image


# - Required Changes  🌵  

- [ ] I made a couple of comments regarding some features you missed, mostly on design flow. Kindly look into it in the comments

# Optional suggestions 

Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing really improves your application.



Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise.

As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

File: app/views/groups/index.html.erb [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+             <div class="col-md-4">
+             <%= image_tag(cat.icon, class: "img-fluid", alt: cat.name, style: "height: 100px; width:100px;") if cat.icon.attached? %>
+        </div>
+            <div class="col-md-8">
+              <div class="card-body text-dark">
+                <div class="d-flex justify-content-between">
+                  <h5 class="card-title"><%= cat.name %></h5>
+                <p class="card-text">Total: $<%= @totals[index] || 0 %></p>
+                </div>
+                <div>
```
Comment (VICTOR-CHUKWU) [inline app/views/groups/index.html.erb]: - [ ] as the previous reviewer pointed out, kindly keep to design guide i.e color, font and spacing, your texts have text decoration which is not on the design guide, also in the total, the text `total` is not in the design guide, so remove it
your design :point_down: 
image
design guide :point_down: 
image

File: app/views/spendings/index.html.erb (hunk 2 above)
Comment (VICTOR-CHUKWU) [reply app/views/spendings/index.html.erb]: - [ ] After adding spending, the page changes and does not look like the given wireframe any longer, kindly fix this
Screenshot from 2023-11-25 14-34-31

**Review by VICTOR-CHUKWU — CHANGES_REQUESTED** [review]
Hi @arnoldnekemiah , 
## STATUS: CHANGES REQUIRED :recycle: 
Great work on making the changes 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


## Suggested changes
_Check the comments under the review._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

File: app/views/spendings/index.html.erb (hunk 2 above)
Comment (KayLemba) [inline app/views/spendings/index.html.erb]: As requested by the previous code reviewer kindly make sure your transaction page matches the given wireframe design. When I add two transactions to a given category the create new transaction button disappears.

This is how your transaction page looks:

Screenshot 2023-11-25 at 7 26 16 PM


Your transaction page should look like this:
Screenshot 2023-11-25 at 7 34 39 PM

Kindly pay attention to detail

**Review by KayLemba — CHANGES_REQUESTED** [review]
Hi @arnoldnekemiah   👋
Good job so far working on the project requirements!
There are some issues that you still need to work on to go to the next project but you are almost there!

To Highlight 🎉

- [x] No linter errors. ✔️

- [x] Git flow is followed. ✔️

- [x] No linter errors

- [x] followed some requirements 
- [x] Tests are passing

Required Changes ♻️
Check the comments under the review in the inline comments.
 
Optional suggestions
Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better.

You can also consider:
N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise.

As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

**Review by Davidosky007 — APPROVED** [review]
Hi @arnoldnekemiah 👋, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights
* All tests passed 👍 
* Well-written PR 👍 
* Good looking readme 👍 
## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**